### PR TITLE
ADD sub-opt prox-TV exp

### PR DESCRIPTION
**Pull request by hcherkaoui**

### CODE REVIEW

File: carpet/metrics.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+    if not hasattr(network, 'training_loss_'):
+        warnings.warn("network seems to not have been trained "
+                      "training_loss_ attribute missing")
+
+    def tv_loss(x, u, lbda):
+        """ TV reg. loss function for Numpy variables. """
+        n_samples = u.shape[0]
+        data_term = 0.5 * np.sum(np.square(u - x))
+        reg = lbda * np.sum(np.abs(np.cumsum(u)))
```
Comment (tomMoral) [inline carpet/metrics.py]: I though `u` is the primal variable for the analysis so this should be a diff no?

```suggestion
        reg = lbda * np.sum(np.abs(np.diff(u)))
```

File: examples/sub_optim_prox.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+    ###########################################################################
+    # Main experiment
+    methods = [  # can be commented in one #
+        ('Analysis TV-Original', 'origtv',
+         dict(prox_tv_max_iter=20), 'tab:orange', '', '--'),
+        # ('Analysis TV-Original', 'origtv',
+        #  dict(prox_tv_max_iter=200), 'tab:orange', '', 'solid'),
+        ('Analysis untrained-TV-Original', 'untrained-origtv',
+         dict(prox_tv_max_iter=20), 'tab:blue', '', '--'),
```
Comment (tomMoral) [inline examples/sub_optim_prox.py]: This parameter is unused.

File: carpet/metrics.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        return (data_term + reg) / n_samples
+
+    if network.verbose > 0:
+        warnings.warn("For convenience network verbose force to 0.")
+        network.verbose = 0
+
+    x = check_tensor(x, device=network.device)
+
+    _, u, _ = init_vuz(network.A, network.D, x, lbda, inv_A=network.inv_A_,
+                       device='cpu')
```
Comment (tomMoral) [inline carpet/metrics.py]:
```suggestion
                       device=network.device)
```

File: carpet/metrics.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        # retrieve parameters
+        mul_lbda = layer_params.get('threshold', 1.0 / network.l_)
+        Wx = layer_params['Wx']
+        Wu = layer_params['Wu']
+        # apply one 'iteration'
+        u = u.matmul(Wu) + x.matmul(Wx)
+        # approx prox-tv
+        approx_prox_z = network.prox_tv(x=u, lbda=lbda * mul_lbda)
+        approx_prox_u = torch.cumsum(approx_prox_z, dim=1)
+        approx_prox_u_npy = approx_prox_u.detach().numpy()
```
Comment (tomMoral) [inline carpet/metrics.py]:
```suggestion
        approx_prox_u_npy = approx_prox_u.detach().cpu().numpy()
```

File: carpet/metrics.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        # true prox-tv
+        u_npy = u.detach().numpy()
```
Comment (tomMoral) [inline carpet/metrics.py]:
```suggestion
        u_npy = u.detach().cpu().numpy()
```

File: carpet/metrics.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        lbda_npy = float(lbda * mul_lbda)
+        prox_u_npy = np.array([prox_tv.tv1_1d(u_, lbda_npy) for u_ in u_npy])
+        # quantify sub-optimality of the approx-prox
+        diff_loss = (tv_loss(u_npy, approx_prox_u_npy, lbda_npy) -  # approx
+            tv_loss(u_npy, prox_u_npy, lbda_npy)                    # true
+            )
```
Comment (tomMoral) [inline carpet/metrics.py]:
```python
diff_loss = (
    tv_loss(u_npy, approx_prox_u_npy, lbda_npy) -  # approx
    tv_loss(u_npy, prox_u_npy, lbda_npy)           # true
)
```

Or 
```python
diff_loss = (tv_loss(u_npy, approx_prox_u_npy, lbda_npy) -  # approx
            tv_loss(u_npy, prox_u_npy, lbda_npy))          # true
```

**Review by tomMoral — COMMENTED** [review]
Some comments. It seems that the cost is not the right one no?

**Comment by hcherkaoui** [comment]
All remarks were taken in account. Merge it

File: examples/sub_optim_prox.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+        l_diff_loss = []
+        for name, type_, kwargs, _, _, _ in methods:
+            print(f"[main script] running {name}")
+            print("-" * 80)
+
+            algo_type = 'origtv' if ('untrained' in type_) else type_
+            network = LearnTVAlgo(algo_type=algo_type, A=A, n_layers=n_layers,
+                                  max_iter=args.max_iter, device=device,
+                                  n_inner_layer=kwargs['n_inner_layer'],
```
Comment (tomMoral) [inline examples/sub_optim_prox.py:104]: why don't use `**kwargs`?
this way your would be agnostic to the class and arguments

**Pull request merged**